frontend/deb/template_rules.go

```diff
@@ -0,0 +1,160 @@
+	return rulesTmpl.Execute(w, &rulesWrapper{spec})
+}
+
+type rulesWrapper struct {
+	*dalec.Spec
+}
+
+func (w *rulesWrapper) Envs() fmt.Stringer {
+	b := &strings.Builder{}
+
+	for k, v := range w.Spec.Build.Env {
+		fmt.Fprintf(b, "export %s := %s\n", k, v)
+	}
+
+	if w.Spec.HasGomods() {
+		fmt.Fprintf(b, "export %s := $(PWD)/%s\n", "GOMODCACHE", gomodsName)
+	}
+
+	return b
+}
+
+func (w *rulesWrapper) OverridePerms() fmt.Stringer {
+	b := &strings.Builder{}
+
+	var fixPerms bool
+	for _, cfg := range w.Spec.Artifacts.Directories.GetConfig() {
+		if cfg.Mode != 0 {
+			fixPerms = true
+			break
+		}
+	}
+
+	if !fixPerms {
+		for _, cfg := range w.Spec.Artifacts.Directories.GetState() {
+			if cfg.Mode != 0 {
+				fixPerms = true
+				break
+			}
+		}
+	}
+
+	if fixPerms {
+		b.WriteString("execute_after_dh_fixperms:\n")
+		b.WriteString("\tdebian/dalec/fix_perms.sh\n\n")
+	}
+
+	return b
+}
+
+// groupUnitsByBaseName indexes the provided list by the unit basename.
+// A unit basename is the name of the unit without the suffix (e.g. ".service", ".socket", etc).
+// The nested map is key'd on the fully resolved unit name.
+func groupUnitsByBaseName(ls map[string]dalec.SystemdUnitConfig) map[string]map[string]dalec.SystemdUnitConfig {
+	idx := make(map[string]map[string]dalec.SystemdUnitConfig)
+	for k, v := range ls {
+		base, suffix := v.SplitName(k)
+		if idx[base] == nil {
+			idx[base] = make(map[string]dalec.SystemdUnitConfig)
+		}
+		idx[base][base+"."+suffix] = v
+	}
+
+	return idx
+}
+
+func (w *rulesWrapper) OverrideSystemd() (fmt.Stringer, error) {
+	b := &strings.Builder{}
+
+	units := w.Spec.Artifacts.Systemd.GetUnits()
+
+	if len(units) == 0 {
+		return b, nil
+	}
+
+	b.WriteString("override_dh_installsystemd:\n")
+
+	grouped := groupUnitsByBaseName(units)
+	sorted := dalec.SortMapKeys(grouped)
+
+	var includeCustomEnable bool
+	for _, basename := range sorted {
+		grouping := grouped[basename]
+
+		needsCustomEnable := requiresCustomEnable(grouping)
+		if needsCustomEnable {
+			includeCustomEnable = true
+		}
+
+		// dh_installsystemd does not want the suffix of the file, so trim it off
+		// here.
+		// Otherwise it will _silently_ fail, *yay*.
+		// We also need to check if there are multiple units with the same base name
+		// with different `Enable` options set.
+		// `dh_installsystemd` cannot deal with this, in those cases we'll write a
+		// custom postinst/postrm script.
+		//
+		// We also only need to do this once per basename, so we don't need to
+		// itterate over every unit.
+
+		// Get the first key which we'll use to check if the unit is enabled.
+		// Either all units are enabled or not enabled OR we need to do custom enable
+		firstKey := maps.Keys(grouping)[0]
+		enable := grouping[firstKey].Enable
+
+		b.WriteString("\tdh_installsystemd --name=" + basename)
+		if !enable || needsCustomEnable {
+			b.WriteString(" --no-enable")
+		}
+		b.WriteString("\n")
+	}
+
+	if includeCustomEnable {
```

Comment:
Can this just be `if needsCustomEnable { ... }`? Do we need the `includeCustomEnable` flag var too?